plugins/modem-manager/fu-sahara-loader.c

```diff
@@ -0,0 +1,582 @@
+	((struct sahara_packet *)(self->data))->length = GUINT32_TO_LE (len);
+
+	return self;
+}
+
+static GByteArray *
+fu_sahara_loader_compose_hello_response_packet(FuSaharaMode mode)
+{
+	GByteArray *self;
+	guint32 len = 0x30;
+
+	self = g_byte_array_sized_new(len);
+	g_byte_array_set_size (self, len);
+
+	((struct sahara_packet *)(self->data))->command_id = GUINT32_TO_LE (SAHARA_HELLO_RESPONSE_ID);
+	((struct sahara_packet *)(self->data))->length = GUINT32_TO_LE (len);
+
+	((struct sahara_packet *)(self->data))->hello_response.version = GUINT32_TO_LE(SAHARA_VERSION);
+	((struct sahara_packet *)(self->data))->hello_response.version_compatible = GUINT32_TO_LE(SAHARA_VERSION_COMPATIBLE);
+	((struct sahara_packet *)(self->data))->hello_response.status = GUINT32_TO_LE(SAHARA_STATUS_SUCCESS);
+	((struct sahara_packet *)(self->data))->hello_response.mode = GUINT32_TO_LE(SAHARA_MODE_IMAGE_TX_PENDING);
+
+	((struct sahara_packet *)(self->data))->hello_response.reserved[0] = GUINT32_TO_LE(1);
+	((struct sahara_packet *)(self->data))->hello_response.reserved[1] = GUINT32_TO_LE(2);
+	((struct sahara_packet *)(self->data))->hello_response.reserved[2] = GUINT32_TO_LE(3);
+	((struct sahara_packet *)(self->data))->hello_response.reserved[3] = GUINT32_TO_LE(4);
+	((struct sahara_packet *)(self->data))->hello_response.reserved[4] = GUINT32_TO_LE(5);
+	((struct sahara_packet *)(self->data))->hello_response.reserved[5] = GUINT32_TO_LE(6);
+
+	return self;
+}
+
+static GByteArray *
+fu_sahara_loader_compose_done_packet(void)
+{
+	GByteArray *self;
+	guint32 len = 0x08;
+
+	self = g_byte_array_sized_new(len);
+	g_byte_array_set_size (self, len);
+
+	((struct sahara_packet *)(self->data))->command_id = GUINT32_TO_LE (SAHARA_DONE_ID);
+	((struct sahara_packet *)(self->data))->length = GUINT32_TO_LE (len);
+
+	return self;
+}
+
+static gboolean
+fu_sahara_loader_send_reset_packet(FuSaharaLoader *self, GError **error)
+{
+	g_autoptr(GByteArray) rx_packet = NULL;
+	g_autoptr(GByteArray) tx_packet = NULL;
+
+	tx_packet = fu_sahara_loader_compose_reset_packet();
+	if (!fu_sahara_loader_send_packet(self, tx_packet, error)) {
+		g_prefix_error(error, "Failed to send reset packet: ");
+		return FALSE;
+	}
+
+	rx_packet = fu_sahara_loader_qdl_read(self, error);
+	if (rx_packet == NULL || sahara_packet_get_command_id(rx_packet) != SAHARA_RESET_RESPONSE_ID) {
+		g_set_error(error,
+			    G_IO_ERROR,
+			    G_IO_ERROR_FAILED,
+			    "Failed to receive RESET_RESPONSE packet");
+		return FALSE;
+	}
+
+	g_debug("Reset succeeded");
+	return TRUE;
+}
+
+static gboolean
+fu_sahara_loader_wait_hello_rsp(FuSaharaLoader *self, GError **error)
+{
+	g_autoptr(GByteArray) rx_packet = NULL;
+	g_autoptr(GByteArray) tx_packet = NULL;
+
+	rx_packet = fu_sahara_loader_qdl_read(self, error);
+	if (rx_packet == NULL) {
+		g_autoptr(GByteArray) ping = NULL;
+		ping = g_byte_array_sized_new(1);
+		g_byte_array_set_size (ping, 1);
```

Comment:
this is a random value, right?

Reply:
Kind of, yes. Sometimes we need to poke a device with a packet for it to reply. The simplest thing is to send 1 byte of empty payload.